src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Rename/RenamerProjectTreeActionHandler.cs

```diff
@@ -0,0 +1,218 @@
+            }
+
+            if (await IsAutomationFunctionAsync() || node.IsFolder)
+            {
+                // Do not display rename Prompt
+                await base.RenameAsync(context, node, value);
+                return;
+            }
+
+            // Get the list of possible actions to execute
+            string oldName = Path.GetFileNameWithoutExtension(oldFilePath);
+            CodeAnalysis.Project? project = GetCurrentProject();
+            if (project is null)
+            {
+                return;
+            }
+
+            CodeAnalysis.Document? oldDocument = GetDocument(project, oldFilePath);
+            if (oldDocument is null)
+            {
+                return;
+            }
+
+            var documentRenameResult = await CodeAnalysis.Rename.Renamer.RenameDocumentAsync(oldDocument, newFileNameWithExtension);
+
+            bool errorsDetected = false;
+            foreach (var action in documentRenameResult.ApplicableActions)
+            {
+                foreach (var error in action.GetErrors())
+                {
+                    errorsDetected = true;
+                    break;
+                }
+            }
+
+            // Check errors before applying changes
+            if (errorsDetected)
+            {
+                string failureMessage = string.Format(CultureInfo.CurrentCulture, VSResources.CannotApplyRename, oldName);
+                _userNotificationServices.ShowWarning(failureMessage);
+                return;
+            }
+
+            if (_languageServiceDone.ContainsKey((oldFilePath, newFileNameWithExtension)))
+            {
+                return;
+            }
+            _languageServiceDone.Add((oldFilePath, newFileNameWithExtension), new SemaphoreSlim(0, 1));
+
+            // Rename the file
+            await base.RenameAsync(context, node, value);
+
+            //Check if there are any symbols that need to be renamed
+            if (documentRenameResult.ApplicableActions.IsEmpty)
+                return;
+
+            // Ask if the user wants to rename the symbol
+            bool userWantsToRenameSymbol = await CheckUserConfirmation(oldName);
+            if (!userWantsToRenameSymbol)
+                return;
+
+            _threadingService.RunAndForget(async () =>
+            {
+                // TODO - implement PublishAsync() to sync with LanguageService
+                // https://github.com/dotnet/project-system/issues/3425)
+                // await _languageService.PublishAsync(treeVersion);
+                //
+                // Because HandleRename() is called after Roslyn, we can set this 
+                // flag when node.FilePath == HandlRename(FilePath)
+                if (_languageServiceDone.TryGetValue((oldFilePath, newFileNameWithExtension), out SemaphoreSlim semLanguegeServiceDone)) {
+                    await semLanguegeServiceDone.WaitAsync();
+                    _languageServiceDone.Remove((node.FilePath, value));
```

Comment:
The Semaphore needs to be disposed after it is used.

Reply:
done in next commit